vnext/Microsoft.ReactNative.Cxx/TurboModuleProvider.cpp

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "TurboModuleProvider.h"
+
+namespace winrt::Microsoft::ReactNative {
+
+// CallInvoker implementation based on JSDispatcher.
+struct AbiCallInvoker final : facebook::react::CallInvoker {
+  AbiCallInvoker(IReactDispatcher const &jsDispatcher) : m_jsDispatcher(jsDispatcher) {}
+
+  void invokeAsync(std::function<void()> &&func) override {
+    m_jsDispatcher.Post([func = std::move(func)]() { func(); });
+  }
+
+  void invokeSync(std::function<void()> &&func) override {
+    if (m_jsDispatcher.HasThreadAccess()) {
+      func();
+    } else {
+      std::mutex mutex;
+      std::condition_variable cv;
+      bool completed{false};
+      std::exception_ptr ex;
+
+      auto lock = std::unique_lock{mutex};
+
+      m_jsDispatcher.Post([&func, &mutex, &cv, &completed, &ex]() {
+        // Since this method is synchronous, we catch all func exceptions and rethrow them in this thread.
+        try {
+          func();
+        } catch (...) {
+          ex = std::current_exception();
+        }
+
+        auto lock = std::unique_lock{mutex};
+        completed = true;
+        cv.notify_all();
+      });
+
+      cv.wait(lock, [&] { return completed; });
+
+      if (ex) {
+        std::rethrow_exception(ex);
+      }
+    }
+
+    // Since func is passed as an r-value, we want to clean it up in this method as we would move the value.
+    func = nullptr;
```

Comment:
should this be done when an exception is [re]thrown too?

Reply:
Good point! We must move it into a local value instead to be exception-safe.